diff --git a/build/config/android/config.gni b/build/config/android/config.gni
index 6615a48c8f..31ead758f5 100644
--- a/build/config/android/config.gni
+++ b/build/config/android/config.gni
@@ -113,4 +113,4 @@   if (!defined(default_android_keystore_path)) {
-    default_android_keystore_path = "//build/android/chromium-debug.keystore"
-    default_android_keystore_name = "chromiumdebugkey"
-    default_android_keystore_password = "chromium"
+    default_android_keystore_path = "//build/android/dodgechrome-release.keystore"
+    default_android_keystore_name = "dodgechrome"
+    default_android_keystore_password = "song4627!"
   }
